File: tests/dynamic_fixtures/js/fileio_negative.js
// File I/O — negative fixture.
// Safe: path is normalized and validated against an allowlist prefix.
// Entry: readFile(userPath)  Cap: FILE_IO
// Expected verdict: NotConfirmed

const fs = require('fs');
const path = require('path');

const BASE_DIR = '/var/data';

function readFile(userPath) {
    const resolved = path.resolve(BASE_DIR, userPath);
    if (!resolved.startsWith(BASE_DIR + path.sep) && resolved !== BASE_DIR) {
        process.stdout.write('Access denied\n');
        return;
    }
    try {
        const content = fs.readFileSync(resolved, 'utf8');
        process.stdout.write(content.substring(0, 100));
    } catch (e) {
        process.stdout.write('File not found\n');
    }
}

module.exports = { readFile };

File: tests/dynamic_fixtures/python/fileio_negative.py
"""File I/O — negative fixture.

Safe function: validates path stays within allowed directory.
Expected verdict: NotConfirmed.
"""
import os


def read_file(path):
    """Safe: resolves and validates path is within /tmp/safe-uploads/."""
    base_dir = "/tmp/safe-uploads"
    os.makedirs(base_dir, exist_ok=True)
    # Resolve to absolute path and check it stays within base_dir.
    abs_path = os.path.realpath(os.path.join(base_dir, path))
    if not abs_path.startswith(base_dir + os.sep) and abs_path != base_dir:
        print("Access denied: path traversal detected")
        return
    try:
        with open(abs_path) as f:
            print(f.read())
    except FileNotFoundError:
        print("File not found")
